File: fabric/src/main/java/juuxel/adorn/platform/fabric/ItemGroupBridgeFabric.java
package juuxel.adorn.platform.fabric;

import juuxel.adorn.item.group.ItemGroupModifyContext;
import juuxel.adorn.platform.ItemGroupBridge;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup;
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents;
import net.minecraft.item.ItemConvertible;
import net.minecraft.item.ItemGroup;
import net.minecraft.registry.RegistryKey;

import java.util.List;

public final class ItemGroupBridgeFabric implements ItemGroupBridge {
    @Override
    public ItemGroup.Builder builder() {
        return FabricItemGroup.builder();
    }

    @Override
    public void addItems(RegistryKey<ItemGroup> group, Function1<? super ItemGroupModifyContext, Unit> configurator) {
        ItemGroupEvents.modifyEntriesEvent(group).register(entries -> {
            var context = new ItemGroupModifyContext() {
                @Override
                public void add(ItemConvertible item) {
                    entries.add(item);
                }

                @Override
                public void addAfter(ItemConvertible after, List<? extends ItemConvertible> items) {
                    entries.addAfter(after, items.toArray(ItemConvertible[]::new));
                }
            };

            configurator.invoke(context);
        });
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/platform/fabric/ItemGroupBridgeFabric.kt
package juuxel.adorn.platform.fabric

import juuxel.adorn.lib.Registered
import juuxel.adorn.platform.ItemGroupBridge
import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents
import net.minecraft.item.ItemConvertible
import net.minecraft.item.ItemGroup
import net.minecraft.util.Identifier

class ItemGroupBridgeFabric : ItemGroupBridge {
    override fun register(id: Identifier, configurator: ItemGroup.Builder.() -> Unit): Registered<ItemGroup> {
        val itemGroup = FabricItemGroup.builder(id)
            .apply(configurator)
            .build()
        return Registered { itemGroup }
    }

    override fun addItems(group: ItemGroup, configurator: ItemGroupBridge.ItemGroupContext.() -> Unit) {
        ItemGroupEvents.modifyEntriesEvent(group).register { entries ->
            val context = object : ItemGroupBridge.ItemGroupContext {
                override fun add(item: ItemConvertible) {
                    entries.add(item)
                }
            }

            configurator(context)
        }
    }
}
